### DISCUSSION TITLE
How can I configure UseNetNamedPipe without Kestrel adding http://localhost:5000?

### DISCUSSION THREAD

**Discussion by pdc1** [CoreWCF/CoreWCF#1336] [Q&A]
Hello.... I am working to port a .Net Framework `ServiceHost` implementation to .Net 8. (Wow, that `ServiceHost` class did a lot of heavy lifting!)

I need the service to communicate with legacy clients using a `NetNamedPipe` endpoint. I was able to mock up a working example using the CoreWCF samples, but now that I am trying to implement a more minimal server using **_only_** named pipes, I have hit a snag.

It appears that the default WebApplicationBuilder uses Kestrel, which in turn creates a default `http://localhost:5000` port. I do not want any additional ports created, just the named pipe.

My ultimate goal is to implement a Windows Service if that helps.

Is there some way to convince Kestrel that it doesn't need a default URL? I do wonder about using "Web" Application, but I am not sure how to use NetNamedPipe without WebHost. 

Here is my sample code so far, I am open to alternate approaches since I have just started fleshing this out.

Thanks for any advice!

```c#
        public static void Main(string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);
            builder.WebHost.UseNetNamedPipe(options =>
            {
                options.Listen("net.pipe://localhost/MyService.svc");
            });
            builder.Services.AddServiceModelServices();

            var app = builder.Build();
            app.UseServiceModel(serviceBuilder =>
            {
                serviceBuilder.AddService<Service>();
                serviceBuilder.AddServiceEndpoint<Service, IService>(new NetNamedPipeBinding(), "/netpipe");
            });
            app.Run();
        }
```

When I run this, I get:
```text
info: CoreWCF.Channels.NamedPipeTransportManager[0]
      Started accepting new NetNamedPipe connections for uri net.pipe://localhost/MyService.svc
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000         <-- Do Not Want
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: [path removed]
```

**Comment by OlegSharov1** [comment]
This is a good question. Did you figure it out?

**Comment by pdc1** [comment]
The best I have come up with is to give Kestrel a different kind of endpoint to listen on so it won't create a default:
```c#
            builder.WebHost.ConfigureKestrel(options =>
            {
                options.ListenNamedPipe("/localhost/null");
            });
```

This seems slightly better to me than leaving an unnecessary http/s port open, but I'd really like an option to have no default at all. I also wonder whether I need Kestrel at all for a NetNamedPipe, but the method to `UseNetNamedPipe` is off of `WebHostBuilder` and not one of the host/application builders.

My other thought is perhaps using a different webhost ([Web server implementations in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/)), though the options there appear to be IIS or `HTTP.sys`, and the latter appears to have the same default socket behavior.

**Comment by Praveer1981** [comment]
@pdc1  I used your code, but I get the below two errors: any idea what the actual root cause could be?
@OlegSharov1 

Error (active)	CS1061	'ConfigureWebHostBuilder' does not contain a definition for 'UseNetNamedPipe' and no accessible extension method 'UseNetNamedPipe' accepting a first argument of type 'ConfigureWebHostBuilder' could be found (are you missing a using directive or an assembly reference?)

Error (active)	CS0121	The call is ambiguous between the following methods or properties: 'ServiceModelApplicationBuilderExtensions.UseServiceModel(IApplicationBuilder, Action<IServiceBuilder>)' and 'ServiceModelApplicationBuilderExtensions.UseServiceModel(IHost, Action<IServiceBuilder>)'

**Comment by pdc1** [comment]
Do you have `CoreWCF.NetNamedPipe` installed? I believe it adds the `UseNetNamedPipe()` method, and hopefully will resolve the second error as well.

**Reply by Praveer1981** [reply]
Thanks, I could solve the problem

**Reply by apreshel** [reply]
@Praveer1981  did you use casting ( `(app as IApplicationBuilder).UseServiceModel(serviceBuilder =>`) , please kindly share the solution that worked for you (casting worked for me, btw)

**Comment by super-brian** [comment]
I do `.UseKestrel(options => options.ListenAnyIP(0))`

**Reply by mconnew** [reply]
I've recently committed a change to remove this requirement. If you grab the latest dev build, you can use an IHost. Take a look at the new unit tests I added to see how to do it.

**Reply by super-brian** [reply]
I think you meant commit 8209a4c and file src/CoreWCF.NetNamedPipe/tests/HostBuilderSetupTests.cs right? 
```
public async Task NetPipeRequestReplyEchoString()
    {
        string testString = new string('a', 3000);
        var host = ServiceHelper.CreateHostBuilder<Startup>(_output, $"{nameof(HostBuilderSetupTests)}/{nameof(NetPipeRequestReplyEchoString)}").Build();
        host.ConfigureUsingStartup<Startup>();

        using (host)
        {
            await host.StartAsync();
            var binding = new System.ServiceModel.NetNamedPipeBinding();
            var factory = new System.ServiceModel.ChannelFactory<IEchoService>(binding,
                new System.ServiceModel.EndpointAddress(new Uri($"net.pipe://localhost/{nameof(HostBuilderSetupTests)}/{nameof(NetPipeRequestReplyEchoString)}/netpipe.svc")));
            var channel = factory.CreateChannel();
            System.ServiceModel.Channels.IChannel ichannel = (System.ServiceModel.Channels.IChannel)channel;
            ichannel.Open();
            var result = channel.EchoString(testString);
            Assert.Equal(testString, result);
            ichannel.Close();
        }
    }
    
     internal class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddServiceModelServices();
        }

        public void Configure(IHost host)
        {
            host.UseServiceModel(builder =>
            {
                builder.AddService<EchoService>();
                builder.AddServiceEndpoint<EchoService, IEchoService>(new CoreWCF.NetNamedPipeBinding(), "netpipe.svc");
            });
        }
    }
    ```

**Reply by super-brian** [reply]
but it is not easy to get the dev build. can you provide me with a link for that? Or can you just public a new version (preview is ok) of CoreWCF.NetNamedPipe? currently it is preview 38 but something like preview 39?

**Reply by mconnew** [reply]
There's instructions in the repo root README.md file, the one which automatically rendered when you visit the main repos page. It details the nuget feed to add to your nuget.config file.

**Reply by mconnew** [reply]
The specific code I was referring to can be found here:
https://github.com/CoreWCF/CoreWCF/blob/8209a4cc8798d4417dbdaa5e904d26294d723a0d/src/CoreWCF.NetNamedPipe/tests/Helpers/ServiceHelper.cs#L37-L52
You can see this test helper just uses Host.CreateDefaultBuilder() to create an HTTP-less host.

**Reply by super-brian** [reply]
Yes, I see that, but NamedPipeWebHostBuilderExtensions class has IHostBuilder UserNetNamedPipe(this IHostBuilder hostbuilder) since the commit 8209a4c (Sep 8 2025), and this commit is after the preview 38 (Jul 24, 2025, 15070b1), so using nuget preview 38, I can't use that. That's why I needed preview 39. (I can't just make those extension methods since it uses internal classes like NetNamedPipeHostedService)

**Comment by apreshel** [comment]
> Hello.... I am working to port a .Net Framework `ServiceHost` implementation to .Net 8. (Wow, that `ServiceHost` class did a lot of heavy lifting!)
> 
> I need the service to communicate with legacy clients using a `NetNamedPipe` endpoint. I was able to mock up a working example using the CoreWCF samples, but now that I am trying to implement a more minimal server using **_only_** named pipes, I have hit a snag.
> 
> It appears that the default WebApplicationBuilder uses Kestrel, which in turn creates a default `http://localhost:5000` port. I do not want any additional ports created, just the named pipe.
> 
> My ultimate goal is to implement a Windows Service if that helps.
> 
> Is there some way to convince Kestrel that it doesn't need a default URL? I do wonder about using "Web" Application, but I am not sure how to use NetNamedPipe without WebHost.
> 
> Here is my sample code so far, I am open to alternate approaches since I have just started fleshing this out.
> 
> Thanks for any advice!
> 
> ```cs
>         public static void Main(string[] args)
>         {
>             var builder = WebApplication.CreateBuilder(args);
>             builder.WebHost.UseNetNamedPipe(options =>
>             {
>                 options.Listen("net.pipe://localhost/MyService.svc");
>             });
>             builder.Services.AddServiceModelServices();
> 
>             var app = builder.Build();
>             app.UseServiceModel(serviceBuilder =>
>             {
>                 serviceBuilder.AddService<Service>();
>                 serviceBuilder.AddServiceEndpoint<Service, IService>(new NetNamedPipeBinding(), "/netpipe");
>             });
>             app.Run();
>         }
> ```
> 
> When I run this, I get:
> 
> ```
> info: CoreWCF.Channels.NamedPipeTransportManager[0]
>       Started accepting new NetNamedPipe connections for uri net.pipe://localhost/MyService.svc
> info: Microsoft.Hosting.Lifetime[14]
>       Now listening on: http://localhost:5000         <-- Do Not Want
> info: Microsoft.Hosting.Lifetime[0]
>       Application started. Press Ctrl+C to shut down.
> info: Microsoft.Hosting.Lifetime[0]
>       Hosting environment: Production
> info: Microsoft.Hosting.Lifetime[0]
>       Content root path: [path removed]
> ```

Do you have any public github account where you develop this project? we be happy to check it out
